Add Documentation after 0.3.0 release (#94)
AbsaOSS/hermes: _docs/0.3/usage/dataset-comparison.md

---
layout: docs
title: Usage - Dataset Comparison
version: '0.3'
categories:
    - '0.3'
    - usage
redirect_from: /docs/usage/dataset-comparison
---

#### Table of contents

- [Table of contents](#table-of-contents)
- [Properties](#properties)
  - [Avaivable Properties](#avaivable-properties)
  - [Default properties are](#default-properties-are)
- [Parameters](#parameters)
- [Usage as library](#usage-as-library)

#### Properties

Properties used in Dataset Comparison can be overriden as standard java opts. For spark that is:

```shell
...
--conf 'spark.driver.extraJavaOptions=-Ddataset-comparison.errColumn=ALFA'
...
```

> ##### Avaivable Properties

Reviewer: ```suggestion
##### Available Properties
```

Author: Fixed

> - `errColumn` is a columns that will be appended in the result parquet file, showing location of the difference found in that row

Reviewer 2: ```suggestion
- `errColumn` is the name of a column that will be appended in the result parquet file, showing location of the difference found in that row
```

> - `actualPrefix` is a prefix of columns comming from the new data source

Reviewer 2: Maybe use **for**? (also bellow)
```suggestion
- `actualPrefix` is a prefix for columns coming from the new data source
```

- `expectedPrefix` is a prefix of columns comming from the referential data source

> - `allowDuplicates` is a switch to allow or disallow duplicate rows in comparisons; duplicates will be ignored

Reviewer 2: _duplicates will be ignored_
What you want to say by that?

Author: Fixed, hopefully.

##### Default properties are

```json
dataset-comparison {
  errColumn = "errCol"
  actualPrefix = "actual"
  expectedPrefix = "expected"
  allowDuplicates = false
}
```

#### Parameters

> Dataset Comparison can recieve two types of parameters + `keys` and `out-path`. There are input general, input specific, `keys` and `out-path`. All should be prefixed with `--` as is standard in command line tools to denote parameter/key followed by its value.

Reviewer 2: This is reads little cumbersome and complicates understanding IMO.
```suggestion
Dataset Comparison can receive the following parameters:
- `keys`
- `out-path`. 
- input general ones
- input specific ones
All should be prefixed with `--` as is standard in command line tools to denote parameter/key followed by its value.
```

Author: well this sentence is not true anymore. Will need more rewriting.

Author: Fixed

> - General parameters are those that will be applied to both reference and new data. These have no prefix

Reviewer 3: ```suggestion
- General parameters will be applied to both reference and new data. These have no prefix.
```

Author: Fixed

> - Specific parameters are those that differentiate if you want to use them for reference or new data source. They are prefixed eaither with `ref-` or `new-`. They override the general ones

Reviewer 3: ```suggestion
- Specific parameters differentiate if you want to use them for the reference or the new data source. They are prefixed either with `ref-` or `new-`. They override the general ones.
```

Author: Fixed

- `keys` is a list of strings delimetered by a `,` (comma) that represent the primary key(s)
- `out-path` is a path to a directory, that will be created by the job, and will hold the diff dataframe and `_METRICS` file

Input parameters (be it general or specific) are governed by the spark datasource formats and options. Imagine having a scala spark code:

```scala
sparkSession
    .read
    .format(<format specified>)
    .options(<map of options provided>)
    .load(<path provided>)
```

> Same way you would provide the Dataset Comparison spark job with `format`, but prefix it with `--`, `path`, but prefix it if either `--ref-` or `--new-`, dependeing on the origing. Maybe add a `--header` option, if you chose `CSV` format or `--rowTag`, if you chose `XML`.

Reviewer 3: ```suggestion
The same way you would provide the Dataset Comparison spark job with `format`, but prefix it with `--`, `path`, but prefix it if either `--ref-` or `--new-`, depending on the origin. Optionally, add a `--header` option, if you chose `CSV` format or `--rowTag`, if you chose `XML`.
```

Reviewer 2: I would again consider not using the personal mood, but stick to imperative as in most of the document.
_The same way provide..._

Author: Fixed

Now how to use it all together:

- `--format`, `--new-format`, `--ref-format` for choosing the format type. This can be `parquet`, `csv`, `xml`, `jdbc`, etc.

> - `--ref-path`, `--new-path` and `--out-path` define paths on HDFS of reference data path, newly created data path and diff output path, repspectively. In case of usage of `jdbc` format these can be replaced by `--ref-dbtable` and `--ned-dbtable`

Reviewer 3: ```suggestion
- `--ref-path`, `--new-path` and `--out-path` define paths on HDFS of reference data path, newly created data path and diff output path, respectively. When using the `jdbc` format, these can be replaced by `--ref-dbtable` and `--ned-dbtable`
```

(btw is it really `--ned-dbtable`, not `--new-dbtable`?)

Author: Fixed

> - `--keys` to specifie the primary key(s) of the table, e.g. `--keys ID1, ID2`. These can be only top level columns.

Reviewer 3: ```suggestion
- `--keys` to specify the primary key(s) of the table, e.g. `--keys ID1, ID2`. These can be only top-level columns.
```

Author: Fixed

- format specific keys prefixed with `--`, so `--header`, `--rowTag`, `--user`, etc.

> **Mandatory fields** are logically the paths/dbtables, format (at least the general one) and any format specific keys that are prescribed by the format used.

Reviewer 2: Humble suggestion: I would leave the _logically_ out.

Author: Fixed

> **NOTE** if you want to use `jdbc` format or others, requiring specific drivers, you should have appropriate drivers on a class path.

Reviewer 3: ```suggestion
**NOTE** if you want to use the `jdbc` format or others, requiring specific drivers, you should make sure that appropriate drivers can be found on the classpath.
```

Author: Fixed

#### Usage as library

For usage as library use class `za.co.absa.hermes.datasetComparison.DatasetComparator`. 

**Prescription of DatasetComparator class**:

```scala
/**
 * Class that is the brain of the DatasetComparison module. This class should be used in case of using DatasetComparison
 * as a library. In case of running the DatasetComparison as SparkJob, please use the DatasetComparisonJob.
 *

>  * @param dataFrameReference is a Dataframe used in a comparison as the origin of truth

Reviewer 2: ```suggestion
 * @param dataFrameReference is the Dataframe used in a comparison as the origin of truth
```

Author: I would like to refuse this because this is one to one copy with the code. Will try and keep in mind

Reviewer 2: How is this (and the other similar) a one-to-one copy with the code? I thought it was a grammatical thing. But happy to ask a native speaker to judge this.

Author: Fixed

>  * @param dataFrameActual is a Dataframe that is being tested

Reviewer 2: ```suggestion
 * @param dataFrameActual is the Dataframe that is being tested
```

Author: I would like to refuse this because this is one to one copy with the code. Will try and keep in mind

Author: Fixed

>  * @param keys is a set of primary keys of the dataset. This highly increases accuracy of the output as we are then able

Reviewer 2: ```suggestion
 * @param keys is a set of field names defining a unique (primary) key of the dataset. This highly increases the accuracy of the output as it makes it possible
```

Author: I would like to refuse this because this is one to one copy with the code.

Author: Fixed

 *             to pinpoint the differences
 * @param config Config object holding project based configurable parameters. Difference to the cliOptions is that these
 *               are meant to stay the same for the project, while cliOptions change for each test
 * @param optionalSchema Optional schema to cherry-pick columns form the two dataframes to compare. For example, if you
 *                       have a timestamp column that will never be the same, you provide a schema without that timestamp
 *                       and it will not be compared.
 * @param sparkSession Implicit spark session.
 */
class DatasetComparator(dataFrameReference: DataFrame,
                        dataFrameActual: DataFrame,
                        keys: Set[String] = Set.empty[String],
                        config: DatasetComparisonConfig = new TypesafeConfig(None),
                        optionalSchema: Option[StructType] = None)
                       (implicit sparkSession: SparkSession)
```

> You create an instance of `DatasetComparator` and then invoke parametrless `compare` method. This method returns `ComparisonResult`.

Reviewer 3: ```suggestion
You create an instance of `DatasetComparator` and then invoke the parameterless `compare` method. This method returns a `ComparisonResult`.
```

Author: Fixed

> You create an instance of `DatasetComparator` and then invoke parametrless `compare` method. This method returns `ComparisonResult`.

Reviewer 2: Again personal vs. imperative mood. The personal is getting in balance with imperative as I go further in the text. 
So up to you, but I would stick to one - whichever you prefer - and keep to it.

Author: Fixed

    .load(<path provided>)
```

In the same way, provide the Dataset Comparison spark job with `format`, but prefix it with `--`, `path`, but prefix it if either `--ref-`, `--new-` or `--out`, depending on the origin and direction. Optionally, add a `--header` option if you chose `csv` format or `--rowTag`, if you chose `XML`.

Now how to use it all together:

- `--format`, `--new-format`, `--ref-format` for choosing the format type. This can be `parquet`, `csv`, `xml`, `jdbc`, etc.
- `--ref-path`, `--new-path`, and `--out-path` define paths on HDFS of the reference data path, newly created data path, and diff output path, respectively. When using the `jdbc` format, these can be replaced by `--ref-dbtable` and `--new-dbtable`

> - `--keys` to specify the primary key(s) of the table, e.g. `--keys ID1, ID2`. These can be only top-level columns.

Reviewer 3: This is very useful, because the general description on how to construct the parameters, albeit being understandable, the directly usable examples are the most easily graspable way of conveying the information to the reader 👍
